### PR TITLE
Flask style Blueprint support

### LINKED ISSUE #40: Blueprint support
Some way of combining similar requests or modulating the application, with their own prefixes, and maybe even middlewares.

Based on Flask's blueprints

### PR DESCRIPTION
**Pull request by The-EDev**
I really don't know what to say here 😅. Everything should be documented in the related issue.

One thing to note is that this is a really large feature, and will most likely have bugs I didn't notice. I tried to fix as many as I can but some are bound to slip through the cracks. So if possible making some manual test to see if everything works as expected would be great.

closes #40

### CODE REVIEW

File: include/crow/app.h [hunk 1]
```diff
@@ -81,6 +84,12 @@ namespace crow
             return router_.new_rule_dynamic(std::move(rule));
         }
 
+//        ///Create a dynamic route for a blueprint using a rule (**Use CROW_BP_ROUTE instead**)
```
Comment (mrozigor) [inline include/crow/app.h:87]: Will be used later or something to delete?

File: include/crow/routing.h [hunk 2]
```diff
@@ -694,13 +708,16 @@ namespace crow
...
     /// A search tree.
     class Trie
     {
     public:
         struct Node
         {
-            unsigned rule_index{};
+            uint16_t rule_index{};
+            // Assign the index to the maximum 32 unsigned integer value by default so that any other number (specifically 0) is a valid BP id.
+            uint16_t blueprint_index{0xFFFF};
```
Comment (mrozigor) [inline include/crow/routing.h:720]: `constexpr uint16_t INVALID_BP_ID = 0xFFFF`?

Comment (The-EDev) [reply include/crow/routing.h:720]: definitely something to delete, my bad..

File: include/crow/app.h (hunk 1 above)
Comment (The-EDev) [reply include/crow/app.h:87]: I forgot to delete this, my bad

File: include/crow/routing.h (hunk 2 above)
Comment (The-EDev) [reply include/crow/routing.h:720]: I'm a little confused, if the value is never going to change, why not use `const`?

File: include/crow/routing.h [hunk 3]
```diff
@@ -1096,20 +1294,85 @@ namespace crow
...
+                blueprints_.emplace_back(&blueprint);
+            }
+            else
+                throw std::runtime_error("blueprint \"" + blueprint.prefix_ + "\" already exists in router");
+        }
+
+        void get_recursive_child_methods(Blueprint* blueprint, std::vector<HTTPMethod>& methods)
+        {
+            //we only need to deal with children if the blueprint has absolutely no methods (meaning its index won't be added to the trie)
+            if (blueprint->static_dir_ == "" && blueprint->all_rules_.empty())
```
Comment (mrozigor) [inline include/crow/routing.h:1322]: `.empty()`?

Comment (mrozigor) [reply include/crow/routing.h:1322]: And in line 1329

File: include/crow/routing.h [hunk 4]
```diff
@@ -1096,20 +1294,85 @@ namespace crow
...
-                            per_methods_[method].trie.add(rule_without_trailing_slash, RULE_SPECIAL_REDIRECT_SLASH);
+                            per_methods_[method].trie.add(rule_without_trailing_slash, RULE_SPECIAL_REDIRECT_SLASH, BP_index != 0xFFFF ? blueprints_[BP_index]->prefix().length() : 0, BP_index);
                         }
                     });
 
         }
 
+        void register_blueprint(Blueprint& blueprint)
+        {
+            if (blueprints_.empty() || std::find(blueprints_.begin(), blueprints_.end(), &blueprint) == blueprints_.end())
```
Comment (mrozigor) [inline include/crow/routing.h:1311]: I think that just `std::find` is enough in this case.

File: include/crow/routing.h [hunk 5]
```diff
@@ -1096,20 +1294,85 @@ namespace crow
...
                         per_methods_[method].rules.emplace_back(ruleObject);
-                        per_methods_[method].trie.add(rule, per_methods_[method].rules.size() - 1);
+                        per_methods_[method].trie.add(rule, per_methods_[method].rules.size() - 1, BP_index != 0xFFFF ? blueprints[BP_index]->prefix().length() : 0, BP_index);
 
                         // directory case:
                         //   request to '/about' url matches '/about/' rule
                         if (has_trailing_slash)
                         {
-                            per_methods_[method].trie.add(rule_without_trailing_slash, RULE_SPECIAL_REDIRECT_SLASH);
+                            per_methods_[method].trie.add(rule_without_trailing_slash, RULE_SPECIAL_REDIRECT_SLASH, BP_index != 0xFFFF ? blueprints_[BP_index]->prefix().length() : 0, BP_index);
```
Comment (mrozigor) [inline include/crow/routing.h:1303]: `INVALID_BP_ID`?

Comment (mrozigor) [reply include/crow/routing.h:1303]: Same in line 1297

File: include/crow/routing.h [hunk 6]
```diff
@@ -1201,6 +1464,43 @@ namespace crow
...
+            {
+                // This statement makes 3 assertions:
+                // 1. The index is above 0.
+                // 2. The index does not lie outside the given blueprint list.
+                // 3. The next blueprint we're adding has a prefix that starts the same as the already added blueprint + a slash (the rest is irrelevant).
+                //
+                // This is done to prevent a blueprint that has a prefix of "bp_prefix2" to be assumed as a child of one that has "bp_prefix".
+                //
+                // If any of the assertions is untrue, we delete the last item added, and continue using the blueprint list of the blueprint found before, the topmost being the router's list
+                if (index > 0 && bp_i[index] < blueprints.size() && blueprints[bp_i[index]]->prefix().substr(0,found_bps[index-1]->prefix().length()+1).compare(std::string(found_bps[index-1]->prefix()+'/')) == 0)
```
[the remark this answers is not in the document]
Comment (The-EDev) [reply include/crow/routing.h:1479]: I know it sounds dumb, But I'm not sure I understand. I guess the second part means move all the checks to a method. but I have no clue what the "Maybe move `index` check to above if" means.. sorry

[the remark this answers is not in the document]
Comment (mrozigor) [reply include/crow/routing.h:1479]: No problem, it means that my explanation was bad :) Also now I see, that my idea was wrong, so please only consider second part - moving all conditions to method.

**Review by mrozigor — APPROVED** [review]

**Pull request merged**